File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int count = 0;
        return helper(root,k,count);
    }
    int helper(TreeNode* root, int k, int& count)
    {
        if(root==NULL) return -1;
        int left = helper(root->left,k,count);
        if(left!=-1) return left;
        count++;
        if(count == k) return root->val;
        int right = helper(root->right,k,count);
        if(right!=-1) return right;
        return -1;
    }
};

File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
      int[]  count = new int[1];
      return Helper(root,k,count);
        
    }
    public int Helper(TreeNode root, int k,int[] count)
    {
        if(root==null) return -1;
        int t = Helper(root.left,k,count);
        if(t!=-1) return t;
        count[0]++;
        if(count[0]==k) return root.val;
        t = Helper(root.right,k,count);
        if(t!=-1) return t;
        return -1;
    }
}